### DISCUSSION TITLE
Gatsby source plugin for Payload CMS - any ideas welcome

### DISCUSSION THREAD

**Discussion by thompsonsj** [payloadcms/payload#2707] [Show and tell]
Hello Payload community! I've just started work on a Gatsby source plugin for Payload CMS.

https://github.com/thompsonsj/gatbsy-source-payload-cms

It's currently working for loading basic data. I've taken care to detect fields where Gatsby should not try to infer a schema, such as rich text, block and upload fields. In these cases, I stringify the JSON to keep it queryable and usable to some extent.

Currently working on:

- Incremental builds (faster build time).
- Gatsby Image CDN support for Gatsby Cloud and Netlify users. See https://support.gatsbyjs.com/hc/en-us/articles/4426393233171-How-to-Enable-Image-CDN. Note that despite the docs saying Gatsby Image CDN isn available on Gatsby Cloud only, the feature is also available on Netlify.
- Asset linking and better parsing of the Payload API response for easier development in Gatsby.

Any feedback/ideas/thoughts are most welcome. For example, Payload CMS (at least by default) does not broadcast field types in API responses. As a result, I am using my own logic to try and determine each field type: https://github.com/thompsonsj/gatbsy-source-payload-cms/blob/main/plugin/src/payload-field-type.ts

I wonder if there is a better way? Have I missed something?

**Comment by jmikrut** [comment]
Hey @thompsonsj — I am not too familiar with Gatsby plugins but I know that @sandren from Candycode is also looking into building a Gatsby plugin for Payload.

Did you see that Payload is able to output its entire GraphQL schema to a text file? 
https://payloadcms.com/docs/graphql/graphql-schema#graphql-schema-generate-script

This could help, but I'm not sure. You also might be able to reuse some of Payload's internal functions like `buildObjectType` which would programmatically allow you to take a Payload collection and then output the exact same GraphQL schema.

https://github.com/payloadcms/payload/blob/master/src/graphql/schema/buildObjectType.ts

Let's keep this conversation going. I think it'd be great to have a Gatsby plugin.

**Reply by thompsonsj** [reply]
This is excellent, thank you. I really like the idea of exporting GraphQL schemas.

I'm going to continue with the REST API approach for now because it is working and the field type detection logic has been battle tested. This will enable me to focus on Image CDN integration with Netlify and incremental builds - which will be useful to work on even if I switch the plugin to use GraphQL.

Once I have time after that I am going to investigate exporting the GraphQL schema for use in Gatsby. Thanks for pointing me in a good direction!
